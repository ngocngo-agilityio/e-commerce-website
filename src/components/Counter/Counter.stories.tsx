import type { Meta, StoryObj } from '@storybook/react';

// Components
import Counter from '.';

const meta = {
  title: 'Counter',
  component: Counter,
  parameters: {
    layout: 'centered',
    nextjs: {
      appDirectory: true,
    },
  },
  tags: ['autodocs'],
} satisfies Meta<typeof Counter>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    initialQuantity: 1,
    onQuantityChange: () => {},
  },
};
