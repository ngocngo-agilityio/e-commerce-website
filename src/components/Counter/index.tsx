'use client';

// Libs
import { useState, useCallback, memo, useMemo } from 'react';
import { Flex, Center, Text, Button } from '@chakra-ui/react';
import { AddIcon, MinusIcon } from '@chakra-ui/icons';

interface Props {
  initialQuantity: number;
  onQuantityChange: (qty: number) => void;
}

const Counter = ({ initialQuantity, onQuantityChange }: Props): JSX.Element => {
  const [quantity, setQuantity] = useState(initialQuantity);

  const isDecrementDisable = useMemo(() => {
    return quantity === 1;
  }, [quantity]);

  const handleChange = useCallback(
    (step: number) => {
      setQuantity((prevQuantity) => {
        const updateCount = prevQuantity + step;

        onQuantityChange(updateCount);

        return updateCount;
      });
    },
    [onQuantityChange],
  );

  const handleDecrement = useCallback(() => {
    handleChange(-1);
  }, [handleChange]);

  const handleIncrement = useCallback(() => {
    handleChange(1);
  }, [handleChange]);

  return (
    <Flex
      justify="space-between"
      w={{ base: '75px', md: '90px' }}
      h={{ base: '28px', md: '34px' }}
      border="1px"
      borderColor="counterBorder"
      borderRadius={{ base: '4px', md: '8px' }}
    >
      <Button
        variant="unstyled"
        minW={5}
        w={{ base: '25px', md: '30px' }}
        h={{ base: '28px', md: '34px' }}
        onClick={handleDecrement}
        isDisabled={isDecrementDisable}
      >
        <MinusIcon boxSize={2} />
      </Button>

      <Center
        w={{ base: '25px', md: '30px' }}
        h={{ base: '28px', md: '34px' }}
        borderInline="1px"
        borderColor="counterBorder"
      >
        <Text
          as="span"
          color="counterText"
          fontWeight="medium"
          fontSize={{ base: 'md', md: 'xl' }}
        >
          {quantity}
        </Text>
      </Center>

      <Button
        minW={5}
        w={{ base: '25px', md: '30px' }}
        h={{ base: '28px', md: '34px' }}
        variant="unstyled"
        onClick={handleIncrement}
      >
        <AddIcon boxSize={2} />
      </Button>
    </Flex>
  );
};

export default memo(Counter);
