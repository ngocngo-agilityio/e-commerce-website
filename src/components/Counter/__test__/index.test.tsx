// Libs
import { fireEvent, render, screen } from '@testing-library/react';

// Components
import Counter from '..';

// Constants
import { COUNTER_DEBOUNCE_TIME } from '@constants';

const mockProps = {
  initialQuantity: 1,
  onQuantityChange: jest.fn(),
};

describe('Counter component', () => {
  test('should match snapshot for Counter', () => {
    const { container } = render(
      <Counter {...mockProps} initialQuantity={2} />,
    );

    expect(container).toMatchSnapshot();
  });

  test('should call onQuantityChange when click increment button', () => {
    jest.useFakeTimers();

    render(<Counter {...mockProps} />);
    const incrementBtn = screen.getByTestId('increment-btn');

    fireEvent.click(incrementBtn);

    jest.advanceTimersByTime(COUNTER_DEBOUNCE_TIME);

    expect(mockProps.onQuantityChange).toHaveBeenCalledWith(2);
    expect(screen.getByText('2')).toBeInTheDocument();
  });

  test('should call onQuantityChange when click decrement button', () => {
    jest.useFakeTimers();

    render(<Counter {...mockProps} initialQuantity={2} />);
    const decrementBtn = screen.getByTestId('minus-btn');

    fireEvent.click(decrementBtn);

    jest.advanceTimersByTime(COUNTER_DEBOUNCE_TIME);

    expect(mockProps.onQuantityChange).toHaveBeenCalledWith(1);
    expect(screen.getByText('1')).toBeInTheDocument();
  });
});
