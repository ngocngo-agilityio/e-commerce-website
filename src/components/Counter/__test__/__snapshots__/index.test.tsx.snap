// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter component should match snapshot for Counter 1`] = `
<div>
  <div
    class="css-6myvwv"
  >
    <button
      aria-label="minus-btn"
      class="chakra-button css-22uwrg"
      data-testid="minus-btn"
      type="button"
    >
      <svg
        class="chakra-icon css-11up8ym"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <g
          fill="currentColor"
        >
          <rect
            height="4"
            width="20"
            x="2"
            y="10"
          />
        </g>
      </svg>
    </button>
    <div
      class="css-ocfx7j"
    >
      <span
        class="chakra-text css-1vec9xb"
      >
        2
      </span>
    </div>
    <button
      aria-label="increment-btn"
      class="chakra-button css-22uwrg"
      data-testid="increment-btn"
      type="button"
    >
      <svg
        class="chakra-icon css-11up8ym"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M0,12a1.5,1.5,0,0,0,1.5,1.5h8.75a.25.25,0,0,1,.25.25V22.5a1.5,1.5,0,0,0,3,0V13.75a.25.25,0,0,1,.25-.25H22.5a1.5,1.5,0,0,0,0-3H13.75a.25.25,0,0,1-.25-.25V1.5a1.5,1.5,0,0,0-3,0v8.75a.25.25,0,0,1-.25.25H1.5A1.5,1.5,0,0,0,0,12Z"
          fill="currentColor"
        />
      </svg>
    </button>
  </div>
</div>
`;
